src/Compilers/Core/Portable/Operations/OperationCache.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Cache data associated with TBoundNode
+    /// </summary>
+    internal class OperationCache<TBoundNode>
+    {
+        // REVIEW: not sure about perf of value tuple used as key on a dictionary. I might need custom type
+        //         if it is somehow not good on perf
+        //
+        // cache's key is (TBoundNode, string) since some of TBoundNode has multiple data associated with it.
```

Comment:
>cache's key is (TBoundNode, string) since some of TBoundNode has multiple data associated with it. [](start = 11, length = 98)

It feels like we should be able to get rid of the need to cache multiple data for the same bound node. We will need a map from a bound node to IOperation node and that node should be able to cache all additional information directly in the IOperation node instance.